---
title: "Bone mineral density data"
output: 
  bookdown::html_document2:
    base_format: rmarkdown::html_vignette
    fig_caption: yes
    toc: false
    number_sections: false
pkgdown:
  as_is: true
bibliography: refs.bib
csl: apa-old-doi-prefix.csl
vignette: >
  %\VignetteIndexEntry{Bone mineral density data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>", echo = TRUE, eval = TRUE, message = FALSE, warning = FALSE
)
```

```{r setup}
library(regcell)
library(tidyverse)
library(janitor)
```

To illustrate the use of CR-Lasso, we consider the bone mineral density (BMD) data from @reppe2010eight. 
The data are publicly available in the [European Bioinformatics Institute Array-Express repository](https://www.ebi.ac.uk/arrayexpress/experiments/E-MEXP-1618).
The BMD data consists of gene expression measurements of 54,675 probes of 84 Norwegian women.

Microarray measurements are often contaminated (noisy), as @rocke2001model highlighted. This contamination can stem from multiple sources, thereby obscuring the gene expression in the data [@zakharkin2005sources].

## Importing the data

Direct links to the data files:

- [Normarrayexpressdata.txt.magetab](https://www.ebi.ac.uk/biostudies/files/E-MEXP-1618/Normarrayexpressdata.txt.magetab)
- [E-MEXP-1618.sdrf.txt](https://www.ebi.ac.uk/biostudies/files/E-MEXP-1618/E-MEXP-1618.sdrf.txt)

```{r, eval = FALSE}
## array expression data
expressiondat = readr::read_tsv("Normarrayexpressdata.txt.magetab") |> 
  t() |> 
  data.frame() |> 
  janitor::row_to_names(row_number = 1) |> 
  tibble::rownames_to_column(var = "id") |> 
  dplyr::mutate(id = readr::parse_number(id)) |> 
  janitor::clean_names() |> 
  dplyr::select(-composite_element_ref) |>
  mutate(across(.cols = -1, .fns = as.numeric)) |> 
  mutate(across(.cols = -1, .fns = log))

# feature data
featuredat = readr::read_tsv("E-MEXP-1618.sdrf.txt") |> 
  janitor::clean_names() |> 
  dplyr::mutate(id = readr::parse_number(source_name)) |> 
  dplyr::select(id,
                hip_t = characteristics_total_hip_t_score) |> 
  arrange(id) # sort obs by patient ID

# merged data
datall = dplyr::left_join(featuredat, expressiondat, 
                          by = "id")
write_rds(datall, file = "datall.rds", compress = "gz")
```

## Feature screening

Given the large number of variables in the dataset, a pre-screening step was implemented to identify the subset of variables that are most correlated with the outcome of interest, the total hip T-score. 
To accomplish this, we first log-transformed all the predictors and then utilized the robust correlation estimate based on Winsorization as in @khan2007robust, instead of the Pearson correlation, since Winsorization is more robust to outliers that may occur in the dataset.
The screened data comprise measurements of $p = 100$ genes from $n = 84$ Norwegian women.

```{r eval = FALSE}
datall = readr::read_rds("datall.rds")
x = datall |> dplyr::select(-id, -hip_t) |> as.matrix() 

y = datall |> dplyr::select(hip_t) |> pull()
# Huber correlation
corh = apply(x, 2,
             FUN = function(xvec){robustHD::corHuber(xvec, y)})

orders = order(abs(corh), decreasing = TRUE)
xscreen = (x[,orders])[,1:100]
# pairs(cbind(y,xscreen[,1:5]))
datascreen = cbind(y, xscreen)
rownames(datascreen) = datall$id
readr::write_rds(datascreen, file = "datascreen.rds")
```

The screened data, `datascreen.rds`, is included in the `regcell` package.

The plot below shows the outlier cell map for 100 screened variables on 84 Norwegian women using the DDC method @rousseeuw_detecting_2018. 
Cells are flagged as outlying if the observed and predicted values differ too much. 
Most cells are blank, showing they are not detected as outliers. 
A red cell means the observed value is significantly higher than the predicted value, and a blue cell means the observed value is significantly lower.

```{r fig-outliervis, fig.height=6, fig.width=10, fig.cap="Outlier cell map for 100 screened variables on 84 Norwegian women. Cells are flagged as outlying if the observed and predicted values differ too much. Most cells are blank, showing they are not detected as outliers. A red cell means the observed value is significantly higher than the predicted value, and a blue cell means the observed value is significantly lower."}
datascreen = read_rds(file = "datascreen.rds")
# data(datascreen)
x = datascreen[,-1]
fit1 <- cellWise::DDC(x)
cellWise::cellMap(
  fit1$stdResid,
  columnlabels = c(1,rep(" ",dim(x)[2]-2),dim(x)[2]),
  rowlabels = c(1,rep(" ",dim(x)[1]-2),dim(x)[1]),
  columnangle = 0,
  colContrast = 1,
  rowtitle = "Observations", columntitle = "Genes",
  sizetitles = 2,
  adjustrowlabels = 0.5,
  adjustcolumnlabels = 0.5)+
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14),
        legend.title = element_text(size=14), #change legend title font size
        legend.text = element_text(size=12)) 

```

```{r}
# Overall average contamination rate
mean((x - fit1$Ximp)!=0)

# Column average
max(colMeans((x - fit1$Ximp)!=0))
which.max(colMeans((x - fit1$Ximp)!=0))

### row average
max(rowMeans((x - fit1$Ximp)!=0))
which.max(rowMeans((x - fit1$Ximp)!=0))
```

On average, the screened genes exhibit an overall contamination rate of `r round(mean((x - fit1$Ximp)!=0),3)*100`%, with probe `r stringr::str_sub(names(which.max(colMeans((x - fit1$Ximp)!=0))),start=2)` having the highest contamination rate of `r round(max(colMeans((x - fit1$Ximp)!=0)),3)*100`%. 
Among the observations, row `r which.max(rowMeans((x - fit1$Ximp)!=0))` in Figure \@ref(fig:fig-outliervis), with patient ID `r names(which.max(rowMeans((x - fit1$Ximp)!=0)))`, shows the highest contamination rate at `r round(max(rowMeans((x - fit1$Ximp)!=0)),3)*100`%.

## All methods compared

```{r, eval = FALSE}
##### methods compared
mtds = list(
  ## spase shooting S, codes arefrom https://github.com/ineswilms/sparse-shooting-S 
  ## package is available from https://github.com/PengSU517/shootings
  # sss = function(y, x){shootings::sparseshooting(x,y)$coef},
  
  ## Rlars, from package robustHD
  rlars = function(y, x){regcell::Rlars(y, x)$betahat}, 
  
  ## MM-Lasso, from https://github.com/esmucler/mmlasso
  # mmlasso = function(y,x){mmlasso::mmlasso(x,y)$coef.MMLasso.ad}, 
  
  ## sparse LTS, from package robustHD
  slts = function(y,x){regcell::SLTS(y,x)$betahat},
  
  ## cellwise regularized Lasso without post regression
  cell_lasso = function(y,x){ 
    fit = regcell::sregcell_std(y = y, x = x, softbeta = TRUE, lambda_zeta = 1, penal = 1, penaldelta =0)
    return(c(fit$intercept_hat, fit$betahat))
  },

  ## cellwise regularized Lasso with post regression
  cell_lasso_post = function(y,x){
    fit = regcell::sregcell_std(y = y, x = x, softbeta = TRUE, lambda_zeta = 1, penal = 1, penaldelta =0)
    return(c(fit$intercept_hat_post, fit$betahat_post))
  },

  ## Lasso, from package glmnet
  lasso  = function(y,x){ 
    return(regcell::lassocv(y,x)$betahat)
  }

)
```


## Analysis with artifically generated response

```{r, eval = FALSE}
library(doParallel)
registerDoParallel(cores=10)
getDoParWorkers()

data(datascreen)
x = as.matrix(robustHD::robStandardize(datascreen[,-1])[,1:100])
ximp = cellWise::DDC(x)$Ximp

result <- foreach(mtd = 1:length(mtds),
                  .packages = c("robustHD", "robustbase" , "mmlasso", "shootings", "cellWise", "regcell"))%:%
  foreach(pr = c(5,10))%:%
  foreach(sigma = c(0.5, 1))%:%
  foreach(m = 1:200)%dopar%{
    set.seed(m)
    error = rnorm(84, sd = sigma)
    beta = sample(c(runif(pr, 1, 1.5),rep(0,100-pr)))

    y = ximp%*%beta + error
    obs = sample(1:84, 17, replace = FALSE)

    ytrain = y[-obs]
    xtrain = x[-obs,]
    ytest = y[obs]
    xtest = ximp[obs,]

    betahat = mtds[[mtd]](ytrain,xtrain)

    tpf<-function(betahat, beta){sum((as.logical(betahat)==as.logical(beta))[which(beta!=0)])}
    tnf<-function(betahat,beta){sum((as.logical(betahat)==as.logical(beta))[which(beta==0)])}
    tp = tpf(betahat[-1], beta)
    tn = tnf(betahat[-1], beta)

    res = ytest - betahat[1] - xtest%*%betahat[-1]
    mape = mean(abs(res))
    rmspe = sqrt(mean(res^2))
    rst = c(m = m, mtd = names(mtds)[mtd], pr = pr, sigma = sigma,
            tp = tp, tn = tn, rmspe = rmspe, mape = mape)

    rst
  }


save(result, file = "result_realdata_artificial.RData")

```


## Leave-one-out Cross-Validation


```{r, eval = FALSE}

data(datascreen)

y = datascreen[,1]
x = as.matrix(robustHD::robStandardize(datascreen[,-1]))
result <- foreach(mtd = 1:length(mtds),
                  .packages = c("robustHD", "robustbase" , "mmlasso","shootings", "cellWise", "regcell"))%:%
  foreach(p = c(100))%:%
  foreach(obs = 1:84)%dopar%{

    set.seed(1)
    x = x[,1:p]
    ytrain = y[-obs]
    xtrain = as.matrix(x[-obs,])
    ytest = y[obs]
    xtest = x[obs,]

    betahat = mtds[[mtd]](ytrain,xtrain)

    res = ytest - betahat[1] - sum(xtest*betahat[-1])
    rst = c(mtd = names(mtds)[mtd], n = obs, p = p, res = res)

    rst
  }


save(result, file = "result_realdata.RData")
```

